manuals/msiecfinfo.1
.Dd June 25, 2026
.Dt MSIECFINFO 1
.Os
.Sh NAME
.Nm msiecfinfo
.Nd determines information about a MSIE Cache File (index.dat)
.Sh SYNOPSIS
.Nm msiecfinfo
.Op Fl c Ar codepage
.Op Fl ahvV
.Ar source
.Sh DESCRIPTION
.Nm msiecfinfo
is a utility to determine information about a MSIE Cache File (index.dat)
.Pp
.Nm msiecfinfo
is part of the
.Nm libmsiecf
package.
.Nm libmsiecf
is a library to access the MSIE Cache File (index.dat) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a
shows allocation information
.It Fl c Ar codepage
codepage of ASCII strings, options: ascii, windows-874, windows-932, \
windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 \
(default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 \
or windows-1258
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# msiecfinfo -a index.dat
msiecfinfo 20260625
.sp
MSIE Cache File information:
	File size:			32768 bytes
	Number of items:		42
	Number of recovered items:	0
.sp
Unallocated data blocks:
	00031232 (0x00007a00) - 00032768 (0x00008000) size: 1536
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr msiecfexport 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libmsiecf/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* address clean up TODOs
* complete thread lock support
** leak
** redirected
** url

* pymsiecf
  - add support for leak: Last cache synchronization date and time
* hash calculation: check what the preceding bytes could be
* use libfdata ?
* msiecfexport:
  - URL what about unknown timestamps
* code clean up
  - add initialize functions
  - rename item values to item value
  - change read values to match libwrc

Tests:
* add cache directory tests

File format
* implement MSIE 3 cache file formats ?
* access records not referred by hash table
  - flag if items are hashed

Performance improvement
* pre allocate item/directory array entries

Tools
* msiecfinfo
  - print cache directories
* msiecfexport
  - export to TSV ?
  - flexible output formatting
  - flexible date/time string formatting

In progress:
* record scan
  - flag entries in hash table as hashed

20140317
* see `git log' for more recent change log
* worked on setup.py

20140131
* 2014 update
* updated dependencies

20131015
* updated dependencies
* updated msvscpp files

20131008
* improved pymsiecf as integer functions

20130929
* worked on setup.py, largely for MSI builds

20120923
* updated msvscpp files

20120922
* updated dependencies

20120806
* fixes for building with stand-alone libuna and libbfio

20120805
* updated dependencies
* bug fix in pymsiecf file object libbfio glue code
* worked on tests

20120529
* updated dependencies

20120528
* worked on pymsiecf simple inheritance of leak, redirected and url items

20120425
* worked on hash calculation

20120423
* added resource file to msvscpp files

20120421
* worked on tests

20120417
* bug fix in record scan due to recent dependency update

20120416
* code clean up

20120414
* textual changes
* worked on tests
* updates and bug fixes in pymsiecf
* fixed codepage 1255 restriction

20120413
* updated dependencies

20120409
* fixed codepage 949 and 950 restriction

20120319
* updated dependencies

20130317
* Small updates

20130310
* worked on url item type support
* msiecfinfo: small changes to output, added major and minor version
* worked on MSIECF 4.7 format support

20130306
* worked on libfole support

20130305
* changes for pymsiecf
* updated libfdatetime
* msiecfexport added expiration and last checked time

20130304
* updated PackageMaker files
* bug fix in pymsiecf FAT date time conversion

20130119
* updated rpm files to include Python bindings

20130114
* updated dpkg files to include Python bindings
* pymsiecf: added generic item fallback object

20130113
* worked on msvscpp files

20130112
* worked on pymsiecf
* fixed issue in spec files

20130111
* worked on pymsiecf

20130110
* removed deprecated functions
* added missing API functions
* worked on pymsiecf

20130109
* 2013 update
* updated dependencies

20120930
* updated dependencies
* changes for libcdata update

20120627
* updated dependencies
* reversed library order in tools for mingw

20120428
* updated dependencies

20120102
* 2012 update

20111120
* updated libfvalue

20111114
* code clean up
* updated array type, list type, offset list
* updated libcstring, libsystem, libfdatetime, libfvalue

20111101
* fix in msiecftools to set codepage

20111031
* updated libcstring, libsystem, libuna, libbfio, libfdatetime, libfvalue

20111030
* worked on UTF8 and UTF16 support
* code clean up
  - added info handle
  - added export handle
* updated msvscpp files

20111029
* updated m4 files
* updated libfdatetime
* added libfvalue
* code clean up
  - removed libmsiecf_item_attach
  - removed libmsiecf_item_detach
* worked on UTF8 and UTF16 support

20111009
* 2011 update
* updated configure.ac and m4 files
* updated include/codepage.h, include/types.h, include/error.h
* updated common
* updated array type, list type, offset list
* updated libcstring, liberror, libnotify, libsystem, libuna, libbfio,
  libfdatetime
* updated pc and spec file
* added gettext support
* updates msvscpp files
* added unused macros
* code clean up
* fixed some leaks in error paths

20101111
* updated configure.ac
* updated include/error.h, include/types.h
* updated common
* updated array type, offset list
* updated liberror, libsystem
* added libmsiecf.rc
* code clean up
* updated msvscpp files

20101023
* added system string support in debug output of url values

20100917
* code clean up
* worked on libcstring support
* worked on offset list changes
* updated log handle
* fix for clean-up bug

20100916
* worked on offset list update
* updated liberror, libbfio

20100912
* email change
* updated configure
* updated include/codepage.h
* updated include/error.h
* updated include/types.h
* added libcstring
* updated common, liberror, libnotify, libsystem, libuna, libbfio,
  libfdatetime 
* updated array type

20100319
* fix in msvscpp files

20100114
* small changes and fixes
* updated check signature
* fix for date time string size for invalid date/time values

20010113
* 2010 update
* updated common, liberror, libnotify, libsystem, libfdatetime, libuna, libbfio
* small changes
* update for libnotify

20091220
* updated libuna
* worked on generic item layer

20091219
* msiecfexport: changed default codepage
* msiecfexport: now used libfdatetime
* library: worked on codepage support
* worked on man pages

20091217
* updated common, libnotify, liberror, libuna, libbfio, libfdatetime
* replaced endian by byte_stream
* worked on generic item layer
  - removed item reference list

20091013
* small adjustments

20091010
* small adjustments

20091009
* worked on msvscpp project files
* updated libmsiecf_notify

20091008
* worked on msvscpp project files
* updated common, libuna, libbfio, liberror, libnotify, libsystem
* updated list and array type

20090912
* worked on file format

20090911
* updated libfdatetime, libbfio, libuna, liberror, libnotify, common
* added libsystem
* added support for m4 directory
* updated spec file and support for empty requires and build requires
* worked on local library support

msiecftools/msiecfinfo.c
/*
 * Shows information obtained from a MSIE Cache File (index.dat).
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "msiecftools_getopt.h"
#include "msiecftools_libcerror.h"
#include "msiecftools_libclocale.h"
#include "msiecftools_libcnotify.h"
#include "msiecftools_libmsiecf.h"
#include "msiecftools_output.h"
#include "msiecftools_signal.h"
#include "msiecftools_unused.h"

info_handle_t *msiecfinfo_info_handle = NULL;
int msiecfinfo_abort                  = 0;

/* Signal handler for msiecfinfo
 */
void msiecfinfo_signal_handler(
      msiecftools_signal_t signal MSIECFTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "msiecfinfo_signal_handler";

	MSIECFTOOLS_UNREFERENCED_PARAMETER( signal )

	msiecfinfo_abort = 1;

	if( msiecfinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     msiecfinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use msiecfinfo to determine information about a MSIE Cache File (index.dat).";

	msiecftools_option_t options[ ] = {
		{ 'a', NULL, "shows allocation information" },
		{ 'c', "codepage", "codepage of ASCII strings, options: ascii, windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258" },
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libmsiecf_error_t *error            = NULL;
	system_character_t *option_codepage = NULL;
	system_character_t *source          = NULL;
	char *program                       = "msiecfinfo";
	system_integer_t option             = 0;
	int number_of_options               = (int) ( sizeof( options ) / sizeof( msiecftools_option_t ) );
	int result                          = 0;
	int show_allocation_information     = 0;
	int verbose                         = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "msiecftools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( msiecftools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	msiecftools_output_version_fprint(
	 stdout,
	 program );

	if( msiecftools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = msiecftools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				msiecftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'a':
				show_allocation_information = 1;

				break;

			case (system_integer_t) 'c':
				option_codepage = optarg;

				break;

			case (system_integer_t) 'h':
				msiecftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				msiecftools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		msiecftools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libmsiecf_notify_set_stream(
	 stderr,
	 NULL );
	libmsiecf_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &msiecfinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( option_codepage != NULL )
	{
		result = info_handle_set_ascii_codepage(
		          msiecfinfo_info_handle,
		          option_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in info handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
	if( info_handle_open_input(
	     msiecfinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file.\n" );

		goto on_error;
	}
	if( info_handle_file_fprint(
	     msiecfinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file information.\n" );

		goto on_error;
	}
	if( show_allocation_information != 0 )
	{
		if( info_handle_unallocated_blocks_fprint(
		     msiecfinfo_info_handle,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to print file unallocated blocks.\n" );

			goto on_error;
		}
	}
	if( info_handle_close_input(
	     msiecfinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &msiecfinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( msiecfinfo_info_handle != NULL )
	{
		info_handle_free(
		 &msiecfinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

